### DISCUSSION TITLE
Request response time

### DISCUSSION THREAD

**Discussion by diegostf** [eModbus/eModbus#103] [General]
I'm starting to use the library (Synchronous TCP) and I have a question about the response time. My network consists of an ESP32 Client, an ESP32 Server 1 and an ATV900 (wired Server 2). All connected to the same router. 
Requests to the ESP32 Server 1 take between 300-800ms to return and requests to the ATV900 take between 25-80ms. 
Is this difference normal? Is it possible to decrease ESP32 Server response time?

**Comment by Miq1** [comment]
I went to the documentation to check the TCP server startup parameters and noticed the whole set up and config stuff is missing there ☹️ 

The server's ``start`` call takes arguments for port number, number of concurrent clients served and a timeout value. This timeout is causing a TCP connection to be closed if no further request has been received within that time. I can imagine that you are experiencing longer response times due to the fact that the timeout has struck and a new connection has to be established first before the request can be served. 

I would suggest you could have a higher timeout value - like 120000 for two minutes - and try again if the response times are getting shorter.

**Comment by diegostf** [comment]
I tried with a longer start time. The result was better at times.

Initialization is:

`MBserver.registerWorker(MY_SERVER, READ_HOLD_REGISTER, &FC03);

MBserver.start(502, 2, 120000);`


![Clipboard01](https://user-images.githubusercontent.com/7763609/131710456-03b713f6-4a65-4794-adda-3cb5062dfca9.png)

**Comment by Miq1** [comment]
How about running the server on a different core? There is a fourth argument to ``start`` that may be 0 or 1 to force the server on that core.
Is your main sketch very busy and may block the server task sometimes?

**Reply by diegostf** [reply]
I tried it with argument 1 to force the server on that core. The weather really got better.
Main loop is not very busy. Most of the time is in the delay.


![Clipboard02](https://user-images.githubusercontent.com/7763609/131712496-7af2e15d-005f-4be0-b867-15e51f8965e1.png)

**Comment by Miq1** [comment]
I would be interested to see the VERBOSE logs for both client and server to see where the time is spent. You either can add ``-DLOG_LEVEL=6`` as compiler option or change the respective setting in Logging.h.

**Reply by diegostf** [reply]
Client:

IP address: 10.0.0.131
[D] 4641| ModbusClientTCP.cpp  [  58] begin: TCP client worker Modbus01TCP started
[D] 4641| ModbusClientTCP.cpp  [  74] setTarget: Target set: 10.0.0.7:502
[V] 4646| ModbusMessage.cpp    [ 505] checkData: Check data #3
[D] 4651| ModbusClientTCP.cpp  [ 100] addToQueue: Queue size: 0
[V] Enqueue: @3FFBD884/6:
  | 0000: 01 03 00 02 00 02                                 |......          |
[D] 4666| ModbusClientTCP.cpp  [ 133] handleConnection: Got request from queue
[D] 7666| ModbusClientTCP.cpp  [  74] setTarget: Target set: 10.0.0.7:502
[V] 7666| ModbusMessage.cpp    [ 505] checkData: Check data #3
[D] 7667| ModbusClientTCP.cpp  [ 100] addToQueue: Queue size: 1
[V] Enqueue: @3FFCCFE4/6:
  | 0000: 01 03 00 02 00 02                                 |......          |
[E][WiFiClient.cpp:258] connect(): socket error on fd 54, errno: 104, "Connection reset by peer"
[D] 10483| ModbusClientTCP.cpp  [ 158] handleConnection: Target connect (10.0.0.7:502).
[D] 10498| ModbusClientTCP.cpp  [ 205] handleConnection: Retry on connect failure...
[D] 10499| ModbusClientTCP.cpp  [ 133] handleConnection: Got request from queue
[D] 10682| ModbusClientTCP.cpp  [  74] setTarget: Target set: 10.0.0.7:502
[V] 10682| ModbusMessage.cpp    [ 505] checkData: Check data #3
[D] 10683| ModbusClientTCP.cpp  [ 100] addToQueue: Queue size: 2
[V] Enqueue: @3FFCCEF8/6:
  | 0000: 01 03 00 02 00 02                                 |......          |
[E][WiFiClient.cpp:258] connect(): socket error on fd 55, errno: 104, "Connection reset by peer"
[D] 10986| ModbusClientTCP.cpp  [ 158] handleConnection: Target connect (10.0.0.7:502).
[D] 11000| ModbusClientTCP.cpp  [ 205] handleConnection: Retry on connect failure...
[D] 11001| ModbusClientTCP.cpp  [ 133] handleConnection: Got request from queue
[E][WiFiClient.cpp:258] connect(): socket error on fd 56, errno: 104, "Connection reset by peer"
[D] 11190| ModbusClientTCP.cpp  [ 158] handleConnection: Target connect (10.0.0.7:502).
Erro: EA - IP connection failed
[D] 11198| ModbusClientTCP.cpp  [ 225] handleConnection: Request popped from queue.
[D] 11205| ModbusClientTCP.cpp  [ 133] handleConnection: Got request from queue
[D] 11396| ModbusClientTCP.cpp  [ 158] handleConnection: Target connect (10.0.0.7:502).
[D] 11397| ModbusClientTCP.cpp  [ 164] handleConnection: Is connected. Send request.
[V] Request packet: @3FFCCD28/12:
  | 0000: 00 01 00 00 00 06 01 03  00 02 00 02              |............    |
[D] 11626| ModbusClientTCP.cpp  [ 275] receive: Received response.
[V] Response packet: @3FFCC740/13:
  | 0000: 00 01 00 00 00 07 01 03  04 02 B8 02 C2           |.............   |
[D] 11632| ModbusClientTCP.cpp  [ 174] handleConnection: Data response.

ID:1, Token:1DF2, Payload: 01 03 04 02 B8 02 C2 - 960ms[D] 11642| ModbusClientTCP.cpp  [ 225] handleConnection: Request popped from queue.
[D] 11650| ModbusClientTCP.cpp  [ 133] handleConnection: Got request from queue
[D] 11660| ModbusClientTCP.cpp  [ 164] handleConnection: Is connected. Send request.
[V] Request packet: @3FFCCD3C/12:
  | 0000: 00 02 00 00 00 06 01 03  00 02 00 02              |............    |
[D] 11907| ModbusClientTCP.cpp  [ 275] receive: Received response.
[V] Response packet: @3FFCC740/13:
  | 0000: 00 02 00 00 00 07 01 03  04 00 58 00 62           |..........X.b   |
[D] 11913| ModbusClientTCP.cpp  [ 174] handleConnection: Data response.

ID:1, Token:29BA, Payload: 01 03 04 00 58 00 62 - 1241ms[D] 11924| ModbusClientTCP.cpp  [ 225] handleConnection: Request popped from queue.
[D] 13698| ModbusClientTCP.cpp  [  74] setTarget: Target set: 10.0.0.7:502
[V] 13698| ModbusMessage.cpp    [ 505] checkData: Check data #3
[D] 13699| ModbusClientTCP.cpp  [ 100] addToQueue: Queue size: 0
[V] Enqueue: @3FFCCD3C/6:
  | 0000: 01 03 00 02 00 02                                 |......          |
[D] 13714| ModbusClientTCP.cpp  [ 133] handleConnection: Got request from queue
[D] 13721| ModbusClientTCP.cpp  [ 164] handleConnection: Is connected. Send request.
[V] Request packet: @3FFCCF30/12:
  | 0000: 00 03 00 00 00 06 01 03  00 02 00 02              |............    |
[D] 14058| ModbusClientTCP.cpp  [ 275] receive: Received response.
[V] Response packet: @3FFCC740/13:
  | 0000: 00 03 00 00 00 07 01 03  04 00 0E 00 18           |.............   |
[D] 14064| ModbusClientTCP.cpp  [ 174] handleConnection: Data response.

ID:1, Token:3582, Payload: 01 03 04 00 0E 00 18 - 376ms[D] 14074| ModbusClientTCP.cpp  [ 225] handleConnection: Request popped from queue.
[D] 16714| ModbusClientTCP.cpp  [  74] setTarget: Target set: 10.0.0.7:502
[V] 16714| ModbusMessage.cpp    [ 505] checkData: Check data #3
[D] 16715| ModbusClientTCP.cpp  [ 100] addToQueue: Queue size: 0
[V] Enqueue: @3FFCCD3C/6:
  | 0000: 01 03 00 02 00 02                                 |......          |
[D] 16730| ModbusClientTCP.cpp  [ 133] handleConnection: Got request from queue
[D] 16737| ModbusClientTCP.cpp  [ 164] handleConnection: Is connected. Send request.
[V] Request packet: @3FFCCF30/12:
  | 0000: 00 04 00 00 00 06 01 03  00 02 00 02              |............    |
[D] 17027| ModbusClientTCP.cpp  [ 275] receive: Received response.
[V] Response packet: @3FFCC740/13:
  | 0000: 00 04 00 00 00 07 01 03  04 02 05 02 0F           |.............   |
[D] 17033| ModbusClientTCP.cpp  [ 174] handleConnection: Data response.

ID:1, Token:414A, Payload: 01 03 04 02 05 02 0F - 329ms[D] 17043| ModbusClientTCP.cpp  [ 225] handleConnection: Request popped from queue.
[D] 19730| ModbusClientTCP.cpp  [  74] setTarget: Target set: 10.0.0.7:502
[V] 19730| ModbusMessage.cpp    [ 505] checkData: Check data #3
[D] 19731| ModbusClientTCP.cpp  [ 100] addToQueue: Queue size: 0
[V] Enqueue: @3FFCCD3C/6:
  | 0000: 01 03 00 02 00 02                                 |......          |
[D] 19746| ModbusClientTCP.cpp  [ 133] handleConnection: Got request from queue
[D] 19753| ModbusClientTCP.cpp  [ 164] handleConnection: Is connected. Send request.
[V] Request packet: @3FFCCF30/12:
  | 0000: 00 05 00 00 00 06 01 03  00 02 00 02              |............    |
[D] 20010| ModbusClientTCP.cpp  [ 275] receive: Received response.
[V] Response packet: @3FFCC740/13:
  | 0000: 00 05 00 00 00 07 01 03  04 02 40 02 4A           |..........@.J   |
[D] 20016| ModbusClientTCP.cpp  [ 174] handleConnection: Data response.

ID:1, Token:4D12, Payload: 01 03 04 02 40 02 4A - 296ms[D] 20026| ModbusClientTCP.cpp  [ 225] handleConnection: Request popped from queue.

**Reply by diegostf** [reply]
Server: 

IP address: 10.0.0.7
[D] 4654| ModbusServer.cpp     [  16] registerWorker: Registered worker for 01/03
[D] 4655| ModbusServerTCPtemp.h [ 151] start: Server task MBserve01F6 started (1073526600).
[D] 4927| ModbusServerTCPtemp.h [ 197] accept: Started client 0 task 1073537496
[D] 4927| ModbusServerTCPtemp.h [ 224] serve: Accepted connection - 1 clients running
[D] 4927| ModbusServerTCPtemp.h [ 241] worker: Worker started, timeout=120000
[I][WiFiClient.cpp:517] connected(): Unexpected: RES: 0, ERR: 0
[D] 5044| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 5044| ModbusServerTCPtemp.h [ 290] worker: Data response
[V] Response: @3FFCD028/13:
  | 0000: 00 01 00 00 00 07 01 03  04 02 B8 02 C2           |.............   |
[D] 5229| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 5230| ModbusServerTCPtemp.h [ 290] worker: Data response
[V] Response: @3FFCD028/13:
  | 0000: 00 02 00 00 00 07 01 03  04 00 58 00 62           |..........X.b   |
[D] 7380| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 7380| ModbusServerTCPtemp.h [ 290] worker: Data response
[V] Response: @3FFCD028/13:
  | 0000: 00 03 00 00 00 07 01 03  04 00 0E 00 18           |.............   |
[D] 10349| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 10350| ModbusServerTCPtemp.h [ 290] worker: Data response
[V] Response: @3FFCD028/13:
  | 0000: 00 04 00 00 00 07 01 03  04 02 05 02 0F           |.............   |
[D] 13323| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 13323| ModbusServerTCPtemp.h [ 290] worker: Data response
[V] Response: @3FFCD028/13:
  | 0000: 00 05 00 00 00 07 01 03  04 02 40 02 4A           |..........@.J   |
[D] 16391| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 16391| ModbusServerTCPtemp.h [ 290] worker: Data response
[V] Response: @3FFCD028/13:
  | 0000: 00 06 00 00 00 07 01 03  04 01 BC 01 C6           |.............   |
[D] 19370| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 19371| ModbusServerTCPtemp.h [ 290] worker: Data response
[V] Response: @3FFCD028/13:
  | 0000: 00 07 00 00 00 07 01 03  04 00 15 00 1F           |.............   |
[D] 22433| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 22433| ModbusServerTCPtemp.h [ 290] worker: Data response
[V] Response: @3FFCD0B0/13:
  | 0000: 00 08 00 00 00 07 01 03  04 00 13 00 1D           |.............   |
[D] 25403| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 25403| ModbusServerTCPtemp.h [ 290] worker: Data response
[V] Response: @3FFCD028/13:
  | 0000: 00 09 00 00 00 07 01 03  04 01 86 01 90           |.............   |
[D] 28475| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 28475| ModbusServerTCPtemp.h [ 290] worker: Data response
[V] Response: @3FFCD028/13:
  | 0000: 00 0A 00 00 00 07 01 03  04 00 67 00 71           |..........g.q   |
[D] 31445| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 31445| ModbusServerTCPtemp.h [ 290] worker: Data response
[V] Response: @3FFCD028/13:
  | 0000: 00 0B 00 00 00 07 01 03  04 01 71 01 7B           |..........q.{   |
[D] 34520| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 34520| ModbusServerTCPtemp.h [ 290] worker: Data response
[V] Response: @3FFCD0B0/13:
  | 0000: 00 0C 00 00 00 07 01 03  04 01 CB 01 D5           |.............   |
[D] 37488| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 37488| ModbusServerTCPtemp.h [ 290] worker: Data response
[V] Response: @3FFCD0B0/13:
  | 0000: 00 0D 00 00 00 07 01 03  04 00 72 00 7C           |..........r.|   |
[D] 40456| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 40456| ModbusServerTCPtemp.h [ 290] worker: Data response
[V] Response: @3FFCD0B0/13:
  | 0000: 00 0E 00 00 00 07 01 03  04 00 5A 00 64           |..........Z.d   |

**Comment by Miq1** [comment]
Huh? I would have bet on the server logging the send action as well - it obviously does not. That would have been the timestamp I was particularly interested in...
The requests are sent quickly (every third second?), the response is recognized and worked on on the server as well. So the time can be spent 
- sending the request via TCP or receiving it. This would point to some network issue. 
- processing the response until it gets sent on the server. Here I am missing the relevant log for the time the response is sent. 
- again transmission via TCP back to the client. This would indicate a network problem as the first possible reason

I must admit I never really cared about the response times, as they were all satisfying in my tests. I am using a very stable WiFi, though, so I never had some network-induced effects yet.

**Reply by diegostf** [reply]
Yes, request are every 3 seconds.

I tested it with 2 different routers (2 networks). Same behavior.
I tested it with other equipment (Schneider ATV900), connected to the router by CAT6 cable. Average response time is 30ms. So it doesn't seem to be a problem on the client.

The problem should be server response time or return via WiFi (???).

What response time did you get?

**Comment by Miq1** [comment]
I am not at the PC anymore.
Would you mind adding the line
ˋˋˋ
LOG_D("Response sent.\n");
ˋˋˋ
Just before line 337 in ModbusServerTCPtemp.h, before the HEXDUMP line?

Then we could see in the log the time the response has been sent.

**Reply by diegostf** [reply]
Client:

[D] 236885| ModbusClientTCP.cpp  [  74] setTarget: Target set: 10.0.0.7:502
[V] 236885| ModbusMessage.cpp    [ 505] checkData: Check data #3
[D] 236886| ModbusClientTCP.cpp  [ 100] addToQueue: Queue size: 0
[V] Enqueue: @3FFAFAFC/6:
  | 0000: 01 03 00 02 00 02                                 |......          |
[D] 236901| ModbusClientTCP.cpp  [ 133] handleConnection: Got request from queue
[D] 236908| ModbusClientTCP.cpp  [ 164] handleConnection: Is connected. Send request.
[V] Request packet: @3FFCBE78/12:
  | 0000: 00 4D 00 00 00 06 01 03  00 02 00 02              |.M..........    |
[D] 237138| ModbusClientTCP.cpp  [ 275] receive: Received response.
[V] Response packet: @3FFCD010/13:
  | 0000: 00 4D 00 00 00 07 01 03  04 03 AE 03 B8           |.M...........   |
[D] 237144| ModbusClientTCP.cpp  [ 174] handleConnection: Data response.

ID:1, Token:39D55, Payload: 01 03 04 03 AE 03 B8 - 269ms[D] 237155| ModbusClientTCP.cpp  [ 225] handleConnection: Request popped from queue.
[D] 239901| ModbusClientTCP.cpp  [  74] setTarget: Target set: 10.0.0.7:502
[V] 239901| ModbusMessage.cpp    [ 505] checkData: Check data #3
[D] 239902| ModbusClientTCP.cpp  [ 100] addToQueue: Queue size: 0
[V] Enqueue: @3FFAFAFC/6:
  | 0000: 01 03 00 02 00 02                                 |......          |
[D] 239917| ModbusClientTCP.cpp  [ 133] handleConnection: Got request from queue
[D] 239924| ModbusClientTCP.cpp  [ 164] handleConnection: Is connected. Send request.
[V] Request packet: @3FFCBE78/12:
  | 0000: 00 4E 00 00 00 06 01 03  00 02 00 02              |.N..........    |
[D] 240109| ModbusClientTCP.cpp  [ 275] receive: Received response.
[V] Response packet: @3FFCD010/13:
  | 0000: 00 4E 00 00 00 07 01 03  04 00 7C 00 86           |.N........|..   |
[D] 240115| ModbusClientTCP.cpp  [ 174] handleConnection: Data response.

ID:1, Token:3A91D, Payload: 01 03 04 00 7C 00 86 - 224ms[D] 240126| ModbusClientTCP.cpp  [ 225] handleConnection: Request popped from queue.
[D] 242917| ModbusClientTCP.cpp  [  74] setTarget: Target set: 10.0.0.7:502
[V] 242917| ModbusMessage.cpp    [ 505] checkData: Check data #3
[D] 242918| ModbusClientTCP.cpp  [ 100] addToQueue: Queue size: 0
[V] Enqueue: @3FFAFAFC/6:
  | 0000: 01 03 00 02 00 02                                 |......          |
[D] 242933| ModbusClientTCP.cpp  [ 133] handleConnection: Got request from queue
[D] 242940| ModbusClientTCP.cpp  [ 164] handleConnection: Is connected. Send request.
[V] Request packet: @3FFCBE78/12:
  | 0000: 00 4F 00 00 00 06 01 03  00 02 00 02              |.O..........    |
[D] 243176| ModbusClientTCP.cpp  [ 275] receive: Received response.
[V] Response packet: @3FFCD010/13:
  | 0000: 00 4F 00 00 00 07 01 03  04 02 87 02 91           |.O...........   |
[D] 243182| ModbusClientTCP.cpp  [ 174] handleConnection: Data response.

ID:1, Token:3B4E5, Payload: 01 03 04 02 87 02 91 - 275ms[D] 243193| ModbusClientTCP.cpp  [ 225] handleConnection: Request popped from queue.
[D] 245933| ModbusClientTCP.cpp  [  74] setTarget: Target set: 10.0.0.7:502
[V] 245933| ModbusMessage.cpp    [ 505] checkData: Check data #3
[D] 245934| ModbusClientTCP.cpp  [ 100] addToQueue: Queue size: 0
[V] Enqueue: @3FFAFAFC/6:
  | 0000: 01 03 00 02 00 02                                 |......          |
[D] 245949| ModbusClientTCP.cpp  [ 133] handleConnection: Got request from queue
[D] 245956| ModbusClientTCP.cpp  [ 164] handleConnection: Is connected. Send request.
[V] Request packet: @3FFCBE78/12:
  | 0000: 00 50 00 00 00 06 01 03  00 02 00 02              |.P..........    |
[D] 246149| ModbusClientTCP.cpp  [ 275] receive: Received response.
[V] Response packet: @3FFCD010/13:
  | 0000: 00 50 00 00 00 07 01 03  04 03 2A 03 34           |.P........*.4   |
[D] 246155| ModbusClientTCP.cpp  [ 174] handleConnection: Data response.


------------------------------------------------------------------------------
Server:
[D] 174504| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 174505| ModbusServerTCPtemp.h [ 290] worker: Data response
[D] 174509| ModbusServerTCPtemp.h [ 316] worker: Response sent.
[V] Response: @3FFCCF3C/13:
  | 0000: 00 4E 00 00 00 07 01 03  04 00 7C 00 86           |.N........|..   |
[D] 177474| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 177475| ModbusServerTCPtemp.h [ 290] worker: Data response
[D] 177478| ModbusServerTCPtemp.h [ 316] worker: Response sent.
[V] Response: @3FFCCF3C/13:
  | 0000: 00 4F 00 00 00 07 01 03  04 02 87 02 91           |.O...........   |
[D] 180545| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 180546| ModbusServerTCPtemp.h [ 290] worker: Data response
[D] 180549| ModbusServerTCPtemp.h [ 316] worker: Response sent.
[V] Response: @3FFCCF3C/13:
  | 0000: 00 50 00 00 00 07 01 03  04 03 2A 03 34           |.P........*.4   |
[D] 183516| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 183516| ModbusServerTCPtemp.h [ 290] worker: Data response
[D] 183520| ModbusServerTCPtemp.h [ 316] worker: Response sent.
[V] Response: @3FFCCF3C/13:
  | 0000: 00 51 00 00 00 07 01 03  04 00 C3 00 CD           |.Q...........   |
[D] 186588| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 186588| ModbusServerTCPtemp.h [ 290] worker: Data response
[D] 186592| ModbusServerTCPtemp.h [ 316] worker: Response sent.
[V] Response: @3FFCCF3C/13:
  | 0000: 00 52 00 00 00 07 01 03  04 02 1A 02 24           |.R..........$   |
[D] 189558| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 189558| ModbusServerTCPtemp.h [ 290] worker: Data response
[D] 189562| ModbusServerTCPtemp.h [ 316] worker: Response sent.
[V] Response: @3FFCCF3C/13:
  | 0000: 00 53 00 00 00 07 01 03  04 02 39 02 43           |.S........9.C   |



![Clipboard03](https://user-images.githubusercontent.com/7763609/131730003-36caf634-176d-48e3-abec-5dee2c84cee3.png)

**Comment by Miq1** [comment]
That is what I expected, thank you. 
The server is taking below 10ms to process the response. 
So the problem must lie in the TCP processing. Another LOG_D behind the write() will tell us if the time is spent filling the packet.

**Reply by diegostf** [reply]
Between lines 313 and 314?

**Reply by diegostf** [reply]
![Clipboard01](https://user-images.githubusercontent.com/7763609/131732452-d366b981-9bde-4b9f-8f04-11eb88703a57.png)

**Comment by Miq1** [comment]
My bad. The log line was already behind the write and flush. So the buffer is filled within the 10ms as well.
I am afraid this is pointing to some problem caused by a network setting or conflicting packets in your WiFi..

**Reply by diegostf** [reply]
Is there any way to check/test this? Can you help me?

**Comment by Miq1** [comment]
Difficult without being at your place. Things you could try:
- get a PC Modbus client on a PC in the same network. Send your requests from that PC to the ESP32 server and check the times.
- try to get Wireshark on a PC in the same network or to get a packet dump from your router to see if there are ACK problems.
- set up a standard web server example on the same ESP32. See if page requests are fast or equally slow.
- get another ESP32 for the server and compare times.

**Reply by diegostf** [reply]
I made the tests:
- With PC Modbus Client, ESP32 server response takes between 150ms and 800ms.
- With PC Modbus Server, response to ESP32 Client is between 10ms and 70ms.
- With another ESP32 server, response times are also between 150ms and 800ms.

    Sometimes the response from the ESP32 server is below 80ms.
    To me, response times below 200ms are quite acceptable for ESP32 Server, but 40% of the time the response is above 400ms.
    I will do other tests with Wireshark and standard web server;

**Comment by Miq1** [comment]
Thank you for all your efforts! 👍🏻 

So the effect is repeatable and really depending on the ESP32 *server* TCP handling. The ESP32 client, using the same TCP stack, seems not to be affected.

I am looking forward to see your Wireshark results - either we will see some communication problem or the time must be spent within the TCP stack on the ESP32.
